### PR TITLE
wip decimals different from 18 test

### PR DESCRIPTION
**Pull request by richmanbtc**
changes

- use latest deps/perpdex-contract
- test withdraw consistency
- fix maxWithdraw to use preview
- use openzeppelin Math.min
- move maxX implementation to long token contract
- add ETH support
- use ReentrancyGuard

todo

- test maxWithdraw all cases
  - preview revert case
  - maxTrade < previewTrade case 
- fix decimals tests

### CODE REVIEW

File: test/longToken/decimals.test.ts [hunk 1]
```diff
@@ -0,0 +1,175 @@
...
+                balance: "-10",
+                totalAssets: "0",
+            },
+        ].forEach(test => {
+            it(test.title, async () => {
+                // force contract balance to be deposit
+                await exchange.setAccountInfo(
+                    longToken.address,
+                    {
+                        collateralBalance: parseAssets(test.balance),
```
Comment (richmanbtc) [inline test/longToken/decimals.test.ts:81]: exchange側でのdecimals変換がバイパスされるので、
外部インターフェースのみを使ったインテグレーションテストの方が良い気がします

Comment (takami-ml) [reply test/longToken/decimals.test.ts:81]: Thank you. I agree with that. I will fix this and `totalAssets`

File: contracts/PerpdexTokenBase.sol [hunk 2]
```diff
@@ -2,93 +2,182 @@
...
         return FullMath.mulDiv(assets, supply, totalAssets());
     }
 
     function convertToAssets(uint256 shares) public view override returns (uint256 assets) {
         uint256 supply = totalSupply();
         if (supply == 0) {
-            // TODO: PerpMath?
-            return FullMath.mulDiv(shares, 10**IERC20Metadata(asset).decimals(), 10**decimals());
+            assets = FullMath.mulDiv(shares, 10**IERC20Metadata(asset).decimals(), 10**decimals());
+            return FullMath.mulDiv(assets, IPerpdexMarket(market).getMarkPriceX96(), FixedPoint96.Q96);
```
Comment (richmanbtc) [inline contracts/PerpdexTokenBase.sol:101]: Perhaps the share price is correct here. The problem is that this is not detected by tests

Comment (richmanbtc) [reply contracts/PerpdexTokenBase.sol]: fixed

File: contracts/PerpdexLongToken.sol [hunk 3]
```diff
@@ -2,92 +2,153 @@
...
-        shares = _previewOpenPosition(true, false, assets);
+        shares = _previewTrade(true, false, assets);
     }
 
     function previewRedeem(uint256 shares) external view override returns (uint256 assets) {
-        assets = _previewOpenPosition(true, true, shares);
+        assets = _previewTrade(true, true, shares);
+    }
+
+    function maxDeposit(address) public view override returns (uint256 maxAssets) {
```
Comment (kei49) [inline contracts/PerpdexLongToken.sol:122]: the input value was not used for maxDeposit()?

File: contracts/PerpdexLongToken.sol [hunk 4]
```diff
@@ -2,92 +2,153 @@
...
+        return _maxTrade(false, false);
+    }
+
+    function maxMint(address) public view override returns (uint256 maxShares) {
```
Comment (kei49) [inline contracts/PerpdexLongToken.sol:126]: the input value was not used for maxMint()?

File: contracts/PerpdexLongToken.sol (hunk 3 above)
Comment (kei49) [reply contracts/PerpdexLongToken.sol:122]: following ERC4626 
https://eips.ethereum.org/EIPS/eip-4626

File: contracts/PerpdexLongToken.sol (hunk 4 above)
Comment (kei49) [reply contracts/PerpdexLongToken.sol:126]: https://eips.ethereum.org/EIPS/eip-4626

**Pull request merged**